experimental/framework/tree-react-api/src/useObservation.ts

```diff
@@ -0,0 +1,128 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import * as React from "react";
+
+interface Subscriptions {
+	/**
+	 * If defined, still needs to be called at some point.
+	 * @remarks
+	 * Clear when called.
+	 */
+	unsubscribe?: () => void;
+}
+
+/**
+ * Wrapper around subscriptions to give it an object identity which can be used with FinalizationRegistry.
+ * @remarks
+ * This indirection is need so inner can be provided to finalizationRegistry as the heldValue and avoid having that cause a leak.
+ * @privateRemarks
+ * This is a named class to make looking for leaks of it in heap snapshots easier.
+ */
+class SubscriptionsWrapper {
+	public readonly inner: Subscriptions = {};
+}
+
+/**
+ * Options for {@link useTreeObservations}.
+ * @input
+ * @public
+ */
+export interface ObservationOptions {
+	/**
+	 * Called when the tracked observations are invalidated.
+	 * This is not expected to have production use cases, but it useful for testing and debugging.
+	 */
+	onInvalidation?: () => void;
+}
+
+/**
+ * Custom hook which invalidates a React Component based on changes to what was observed during `trackDuring`.
+ *
+ * @param trackDuring - Called synchronously: can make event subscriptions which call the provided `invalidate` function.
+ * Any such subscriptions should be cleaned up via the returned `unsubscribe` function which will only be invoked if `invalidate` is not called.
+ * If `invalidate` is called, the code calling it should remove any subscriptions before calling it.
+ */
+export function useObservation<TResult>(
+	trackDuring: (invalidate: () => void) => { result: TResult; unsubscribe: () => void },
+	options?: ObservationOptions,
+): TResult {
+	// Use a React state hook to invalidate this component something tracked by `trackDuring` changes.
+	const [subscriptions, setSubscriptions] = React.useState<SubscriptionsWrapper>(
+		new SubscriptionsWrapper(),
+	);
+
+	// Because `subscriptions` is used in `finalizationRegistry` for cleanup, it is important that nothing save a reference to it which is retained by the invalidation callback.
```

Comment:
Thanks for all the docs here! These offer a lot of useful context for what's going on.